Remove `Material` trait from prelude

efikay opened the pull request:
# Objective

- Remove `Material` trait from `bevy::prelude::*` imports to prevent naming conflicts in game dev projects.
- "Material" is a very common term in game development, and this change will help prevent naming collisions in user code.

## Solution

- Removed `Material` trait re-export from `bevy_pbr`'s `prelude` module in order to prevent further re-exports which leads into desired outcome.
- Updated imports to make sure code previously used `Material` via prelude now still accessing it using new import.
- Users who need `Material` trait can now explicitly import it from `bevy::pbr::Material`.

## Testing

- Built a project using modified Bevy version – build passed successfully:
```bash
cargo build
cargo build --features dynamic_linking
cargo build --features bevy/dynamic_linking
```
- Ran all tests in the root directory using `cargo test` – no errors encountered
- Tested on real project – works as expected
- All platforms should behave identically as this is purely an import change

---

## Migration guide

Code that previously relied on prelude import will need to add:

```rust
use bevy::pbr::Material;
```

Example before:

```rust
use bevy::prelude::*;

pub struct MyAwesomeMaterial {
  // ...
}

impl Material for MyAwesomeMaterial {
  // ...
}
```

After:

```rust
use bevy::pbr::Material;
use bevy::prelude::*;

pub struct MyAwesomeMaterial {
  // ...
}

impl Material for MyAwesomeMaterial {
  // ...
}
```

efikay committed:
Remove material::Material from prelude

efikay committed:
Remove Material trait from bevy::prelude::*

efikay committed:
Fix missing bevy::pbr::Material imports

alice-i-cecile reviewed (approved):
Good call. I think that this is rarely enough done that we can safely cut it from the prelude, and I agree that this is a high-traffic name for games :) Nicely done!

efikay committed:
Add migration guide

efikay committed:
Fix migration guide(?)

Found whitespace line missing.
I don't have tool that CI is using so I'm just guessing.

tychedelia reviewed (approved):
If we do this, we should also remove `Material2d` from the prelude for consistency (although it's unlikely to be a conflict).

efikay commented:
Mention me pls if you're all up for it (Material2d from prelude also).

❓: Why my Markdown is not passing CI check? How can I check Markdown by myself? Thank you

IceSentry commented:
I don't think I agree with this. Material is a very specific thing in bevy and if you are using bevy and naming something else Material seems like it can only create more confusion. Material is a very basic type used all over the place for bevy, it's exactly what a prelude is for.

alice-i-cecile commented on release-content/migration-guides/no_material_in_prelude.md:
```diff
@@ -0,0 +1,30 @@
+Code that previously relied on `Material` from `prelude` will need to add:
```
This needs a heading and linked PR number :) See the sibling migration guides for a template to copy-paste.

efikay committed:
Update migration guide with meta-data

alice-i-cecile commented:
@efikay, just so you know, you've done a great job on this PR, and we don't need anything else from you (but feel free to join in on the discussion). However, our Rendering Subject Matter Experts disagree on what should be done here. @tychedelia is in favor, and @IceSentry is against it.

If they cannot come to an agreement, I'll close this out as not planned and preserved the status quo, but we still appreciate you sparking this discussion.

efikay commented:
@alice-i-cecile Thank you. I won't complain if you decide to decline these changes. This whole PR is the subject of taste (and particular knowledge what prelude is all abou if there are any specifics about it)

IceSentry commented:
I'd like to know a situation where you would be using bevy_pbr but also need a custom struct called `Material`. The only case I can think of is for making a custom renderer and in that case you should not be using bevy_pbr anyway. I'm not completely against it, but I'm not sure I understand the scenario where this would be an issue.

alice-i-cecile commented:
`Material` is a fairly reasonable name for gameplay "material": everything from "what is this pickaxe made of" to "what is this diggable tile composed of".

The conflict isn't super important, because it's a trait vs struct, but I'm also not convinced that users are going to be defining more than a handful of custom materials per project, or referencing this trait's methods outside of their specific material files.

IceSentry commented:
I think it's safe to close this for now. If anyone has another opinion feel free to reopen.
